integration-tests/src/test/java/io/datanerds/grpc/MockListener.java

```diff
@@ -0,0 +1,50 @@
+package io.datanerds.grpc;
+
+import io.grpc.Attributes;
+import io.grpc.NameResolver;
+import io.grpc.ResolvedServerInfoGroup;
+import io.grpc.Status;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class MockListener implements NameResolver.Listener {
+    private List<ResolvedServerInfoGroup> servers = Collections.synchronizedList(new ArrayList<>());
+    private Attributes attributes;
+    private List<Status> errors = Collections.synchronizedList(new ArrayList<>());
+
+    @Override
+    public void onUpdate(List<ResolvedServerInfoGroup> servers, Attributes attributes) {
+        synchronized (this) {
```

Comment:
Would probably introduce: `private final Object lock = new Object();` to lock on that instead of this?